### PR TITLE
Event Subscription Filters 

### PR DESCRIPTION
**Pull request by NigeWarren**
Allows a user to define (in the Server) and Specify (in the client, hence protocol) the name (FQCN) of the filter to apply to 

Filter must implement the Predicate function type from the standard library and should use a unique naming system, preferably the Java FQCN convention.

Predicates are added to a Filter Factory in the server, assumed at start up time.

```
 FilterFactory.addFilter(filterName, event -> {
            return event.getInteger("num"); != 7;
        });
```

And specified when a subscription for an event stream is prepared in the client as part of the Subscription Descriptor. 

```
       SubDescriptor descriptor = new SubDescriptor();
        descriptor.setChannel(TEST_CHANNEL_1);
        descriptor.setFilterName(filterName);
        descriptor.setStartEventNum(0);     // replay all the events and filter num 7
```

The use of the Predicates in the server looks fine and supported by tests. I'm not so sure about the use of the not very idiomatic 'Factory' to store the name/predicate pairs.  There may be better (more natural and idiomatic to Java 8)  ways of doing this.

### CODE REVIEW

File: src/main/java/io/mewbase/server/impl/SubscriptionBase.java [hunk 1]
```diff
@@ -81,20 +87,25 @@ public void unsubscribe() {
...
                 return;
             }
         }
-        onReceiveFrame(pos, frame);
+
+        BsonObject event = record.getBsonObject("event");
+        if (subsFilter.test(event)) {
+            onReceiveFrame(pos, record);
+        }
+        return;
```
Comment (purplefox) [inline src/main/java/io/mewbase/server/impl/SubscriptionBase.java]: redundant return

File: src/test/java/io/mewbase/ChannelsTest.java [hunk 2]
```diff
@@ -228,4 +230,49 @@ public void testSubscribeFromTimestamp(TestContext context) throws Exception {
     }
 
 
+    @Test
```
Comment (purplefox) [inline src/test/java/io/mewbase/ChannelsTest.java]: I think we should have a test using an invalid filter name

File: src/main/java/io/mewbase/server/filter/FilterFactory.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+import io.mewbase.bson.BsonObject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
+
+
+public class FilterFactory {
```
Comment (purplefox) [inline src/main/java/io/mewbase/server/filter/FilterFactory.java]: In general, I am not a big fan of using statics for things like this. They pollute the static space and make it hard to have multiple apps with different filters in the same JVM.

I'd prefer to see an approach similar to how what we do with queries used.

Comment (NigeWarren) [inline src/main/java/io/mewbase/server/filter/FilterFactory.java]: Yep agreed - just talk me through the preferred pattern when next in.

Comment (purplefox) [inline src/main/java/io/mewbase/server/filter/FilterFactory.java]: For consistency we could use a similar pattern to how we build queries: https://github.com/Tesco/mewbase/blob/master/src/main/java/io/mewbase/server/QueryBuilder.java

Comment (NigeWarren) [inline src/main/java/io/mewbase/server/filter/FilterFactory.java]: ok

**Pull request merged**